include/nuttx/clk/clk_provider.h

```diff
@@ -0,0 +1,262 @@
+  uint8_t                     enable_count;
+  uint8_t                     flags;
+  uint32_t                    rate;
+  uint32_t                    new_rate;
+  FAR struct clk_s           *new_parent;
+  FAR struct clk_s           *new_child;
+  FAR void                   *private_data;
+  struct list_node            children;
+  struct list_node            node;
+  FAR const char             *parent_names[0];
+};
+
+struct clk_ops_s
+{
+  CODE int       (*enable)(FAR struct clk_s *clk);
+  CODE void      (*disable)(FAR struct clk_s *clk);
+  CODE int       (*is_enabled)(FAR struct clk_s *clk);
+  CODE uint32_t  (*recalc_rate)(FAR struct clk_s *clk, uint32_t parent_rate);
+  CODE uint32_t  (*round_rate)(FAR struct clk_s *clk,
+                               uint32_t rate, uint32_t *parent_rate);
+  CODE uint32_t  (*determine_rate)(FAR struct clk_s *clk, uint32_t rate,
+                                   uint32_t *best_parent_rate,
+                                   struct clk_s **best_parent_clk);
+  CODE int       (*set_parent)(FAR struct clk_s *clk, uint8_t index);
+  CODE uint8_t   (*get_parent)(FAR struct clk_s *clk);
+  CODE int       (*set_rate)(FAR struct clk_s *clk, uint32_t rate,
+                             uint32_t parent_rate);
+  CODE int       (*set_rate_and_parent)(FAR struct clk_s *clk, uint32_t rate,
+                                        uint32_t parent_rate,
+                                        uint8_t index);
+  CODE int       (*get_phase)(FAR struct clk_s *clk);
+  CODE int       (*set_phase)(FAR struct clk_s *clk, int degrees);
+};
+
+struct clk_gate_s
+{
+  uint32_t            reg;
+  uint8_t             bit_idx;
+  uint8_t             flags;
+};
+
+struct clk_fixed_rate_s
+{
+  uint32_t            fixed_rate;
+  uint8_t             flags;
+};
+
+struct clk_fixed_factor_s
+{
+  uint8_t             mult;
+  uint8_t             div;
+};
+
+struct clk_divider_s
+{
+  uint32_t            reg;
+  uint8_t             shift;
+  uint8_t             width;
+  uint16_t            flags;
+};
+
+struct clk_phase_s
+{
+  uint32_t            reg;
+  uint8_t             shift;
+  uint8_t             width;
+  uint8_t             flags;
+};
+
+struct clk_fractional_divider_s
+{
+  uint32_t            reg;
+  uint8_t             mwidth;
+  uint8_t             nwidth;
+  uint8_t             mshift;
+  uint8_t             nshift;
+  uint8_t             flags;
+};
+
+struct clk_multiplier_s
+{
+  uint32_t            reg;
+  uint8_t             shift;
+  uint8_t             width;
+  uint8_t             flags;
+};
+
+struct clk_mux_s
+{
+  uint32_t            reg;
+  uint8_t             width;
+  uint8_t             shift;
+  uint8_t             flags;
+};
+
+/****************************************************************************
+ * Public Function Prototypes
+ ****************************************************************************/
+
+FAR struct clk_s *clk_register(FAR const char *name,
```

Comment:
Where is clk_register() implemented? It is declared here, called in some places but I didn't see it implemented anywhere. It cannot be a Generic selection because we support C11 on NuttX :-)

Reply:
It's here:
https://github.com/apache/incubator-nuttx/pull/4918/files#diff-fa5c6e68e529521e8d1eb81aa3d6fcd208f686db83c964cafb417ecb8bd62832R1132-R1208